contracts/pcv/compound/ERC4626CErc20Wrapper.sol

```diff
@@ -0,0 +1,240 @@
+        return address(cTokenUnderlying);
+    }
+
+    /// @notice Total amount of the underlying asset that
+    /// is "managed" by Vault.
+    function totalAssets() external view returns(uint256) {
+        return cToken.getCash() + cToken.totalBorrows();
+    }
+
+    /*///////////////////////////////////////////////////////////////
+                ERC4626 Deposit/Withdrawal logic
+    //////////////////////////////////////////////////////////////*/
+
+    /// @notice Mints `shares` Vault shares to `receiver` by
+    /// depositing exactly `amount` of underlying tokens.
+    function deposit(uint256 assets, address receiver) external returns(uint256 shares) {
+        // transfer user tokens to self
+        cTokenUnderlying.safeTransferFrom(msg.sender, address(this), assets);
+
+        // mint cTokens
+        cTokenUnderlying.approve(address(cToken), assets);
+        uint256 balanceBefore = cToken.balanceOf(address(this));
+        require(cToken.mint(assets) == 0, "ERC4626CErc20Wrapper: error on cToken.mint");
+        shares = cToken.balanceOf(address(this)) - balanceBefore;
+
+        // mint shares to receiver
+        _mint(receiver, shares);
+
+        // emit event
+        emit Deposit(msg.sender, receiver, assets, shares);
+
+        return shares;
+    }
+
+    /// @notice Mints exactly `shares` Vault shares to `receiver`
+    /// by depositing `amount` of underlying tokens.
+    function mint(uint256 shares, address receiver) external returns(uint256 assets) {
+        // compute exact amount of assets to send
+        uint256 exchangeRate = cToken.exchangeRateCurrent();
+        assets = shares * exchangeRate / 1e18;
+
+        // mint cTokens
+        cTokenUnderlying.safeTransferFrom(msg.sender, address(this), assets);
+        cTokenUnderlying.approve(address(cToken), assets);
+        uint256 balanceBefore = cToken.balanceOf(address(this));
+        require(cToken.mint(assets) == 0, "ERC4626CErc20Wrapper: error on cToken.mint");
+        uint256 actualShares = cToken.balanceOf(address(this)) - balanceBefore;
+        assert(shares == actualShares); // double check no rounding errors
+
+        // mint shares to receiver
+        _mint(receiver, shares);
+
+        // emit event
+        emit Deposit(msg.sender, receiver, assets, shares);
+
+        return assets;
+    }
+
+    /// @notice Redeems `shares` from `owner` and sends `assets`
+    /// of underlying tokens to `receiver`.
+    function withdraw(uint256 assets, address receiver, address owner) external returns(uint256 shares) {
+        // redeem cTokens and get the actual number of shares burnt
+        uint256 balanceBefore = cToken.balanceOf(address(this));
+        require(cToken.redeemUnderlying(assets) == 0, "ERC4626CErc20Wrapper: error on cToken.redeemUnderlying");
+        shares = balanceBefore - cToken.balanceOf(address(this));
+
+        // Check that owner approved spending on behalf of the caller
+        if (msg.sender != owner) {
```

Comment:
Allowance check should be first

Reply:
we don't know the actual number of shares burnt before we redeem, we can only approximate it with the stored exchange rate, and it might be not 100% accurate, which can be a problem if he allowance is a precise number of shares